dev/tools/test/update_engine_version_test.dart

```diff
@@ -0,0 +1,296 @@
+    setUp(() async {
+      await setupRepo(branch: 'master');
+    });
+
+    test('merge-base HEAD upstream/master on non-LUCI when upstream is set', () async {
+      await setupRemote(remote: 'upstream');
+
+      final ProcessRunnerResult mergeBaseHeadUpstream = await processRunner.runProcess(<String>[
+        'git',
+        'merge-base',
+        'HEAD',
+        'upstream/master',
+      ]);
+      await processRunner.runProcess(<String>[testRoot.binInternalUpdateEngineVersion.path]);
+
+      expect(testRoot.binInternalEngineVersion, exists);
+      expect(
+        testRoot.binInternalEngineVersion.readAsStringSync(),
+        equalsIgnoringWhitespace(mergeBaseHeadUpstream.stdout),
+      );
+    });
+
+    test('merge-base HEAD origin/master on non-LUCI when upstream is not set', () async {
+      await setupRemote(remote: 'origin');
+
+      final ProcessRunnerResult mergeBaseHeadOrigin = await processRunner.runProcess(<String>[
+        'git',
+        'merge-base',
+        'HEAD',
+        'origin/master',
+      ]);
+      await processRunner.runProcess(<String>[testRoot.binInternalUpdateEngineVersion.path]);
+
+      expect(testRoot.binInternalEngineVersion, exists);
+      expect(
+        testRoot.binInternalEngineVersion.readAsStringSync(),
+        equalsIgnoringWhitespace(mergeBaseHeadOrigin.stdout),
+      );
+    });
+
+    test('rev-parse HEAD when running on LUCI', () async {
+      environment['LUCI_CONTEXT'] = '_NON_NULL_AND_NON_EMPTY_STRING';
+      await processRunner.runProcess(<String>[testRoot.binInternalUpdateEngineVersion.path]);
+
+      final ProcessRunnerResult revParseHead = await processRunner.runProcess(<String>[
+        'git',
+        'rev-parse',
+        'HEAD',
+      ]);
+      expect(testRoot.binInternalEngineVersion, exists);
+      expect(
+        testRoot.binInternalEngineVersion.readAsStringSync(),
+        equalsIgnoringWhitespace(revParseHead.stdout),
+      );
+    });
+  });
+
+  group('if DEPS or engine/src/.gn are omitted', () {
+    setUp(() {
+      for (final File f in <File>[testRoot.deps, testRoot.engineSrcGn]) {
+        f.createSync(recursive: true);
+      }
+    });
+
+    test('[DEPS] engine.version is blank', () async {
+      testRoot.deps.deleteSync();
+
+      await processRunner.runProcess(<String>[testRoot.binInternalUpdateEngineVersion.path]);
+
+      expect(testRoot.binInternalEngineVersion, exists);
+      expect(testRoot.binInternalEngineVersion.readAsStringSync(), equalsIgnoringWhitespace(''));
+    });
+
+    test('[engine/src/.gn] engine.version is blank', () async {
+      testRoot.engineSrcGn.deleteSync();
+
+      await processRunner.runProcess(<String>[testRoot.binInternalUpdateEngineVersion.path]);
+
+      expect(testRoot.binInternalEngineVersion, exists);
+      expect(testRoot.binInternalEngineVersion.readAsStringSync(), equalsIgnoringWhitespace(''));
+    });
+  });
+}
+
+/// A FrUT, or "Flutter Root"-Under Test (parallel to a SUT, System Under Test).
```

Comment:
Why introduce a new term if it's not used anywhere?

Reply:
It's being used here :)

Reply:
I could call it "FlutteRroot", but it's a private class in a single test and there are two roots.